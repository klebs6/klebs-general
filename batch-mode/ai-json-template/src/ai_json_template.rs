// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}
