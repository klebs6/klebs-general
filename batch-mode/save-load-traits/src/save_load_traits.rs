// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
