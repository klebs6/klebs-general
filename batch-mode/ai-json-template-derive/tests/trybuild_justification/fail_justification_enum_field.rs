// ---------------- [ File: ai-json-template-derive/tests/trybuild_justification/fail_justification_enum_field.rs ]
// tests/trybuild_justification/fail_justification_enum_field.rs
//
// We nest an enum inside the struct => your macro doesn't handle enums in the fields
// => compilation fails.

#![allow(dead_code)]
#![allow(unused_imports)]

use ai_json_template_derive::*;
use ai_json_template::*;
use serde::*;
use getset::*;
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;
use tracing::*;

#[derive(PartialEq,Eq,SaveLoad, Debug, Clone, Serialize, Deserialize)]
enum BadEnum {
    A,
    B,
}

#[derive(PartialEq,Eq,SaveLoad, Debug, Clone, Serialize, Deserialize, Getters, Setters, Builder)]
#[derive(AiJsonTemplate, AiJsonTemplateWithJustification)]
#[getset(get = "pub", set = "pub")]
#[builder(setter(into))]
struct ContainsEnum {
    // This field is an enum => should fail.
    data: BadEnum,
}

fn main() {}
