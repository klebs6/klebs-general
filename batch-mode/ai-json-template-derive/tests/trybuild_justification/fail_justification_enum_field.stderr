error[E0412]: cannot find type `BadEnumJustification` in this scope
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ not found in this scope

error[E0412]: cannot find type `BadEnumJustification` in this scope
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ not found in this scope
   |
help: you might be missing a type parameter
   |
31 | }<BadEnumJustification>
   |  ++++++++++++++++++++++

error[E0412]: cannot find type `BadEnumConfidence` in this scope
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ not found in this scope

error[E0412]: cannot find type `BadEnumConfidence` in this scope
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ not found in this scope
   |
help: you might be missing a type parameter
   |
31 | }<BadEnumConfidence>
   |  +++++++++++++++++++

error[E0277]: the trait bound `BadEnum: ai_json_template::AiJsonTemplate` is not satisfied
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ the trait `ai_json_template::AiJsonTemplate` is not implemented for `BadEnum`
   |
   = help: the following other types implement trait `ai_json_template::AiJsonTemplate`:
             ContainsEnum
             HashMap<K, V>
             Vec<T>
             std::string::String

error[E0277]: the trait bound `ContainsEnum: Default` is not satisfied
  --> tests/trybuild_justification/fail_justification_enum_field.rs:25:26
   |
25 | #[derive(AiJsonTemplate, AiJsonTemplateWithJustification)]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `ContainsEnum`

error[E0369]: binary operation `==` cannot be applied to type `ContainsEnum`
  --> tests/trybuild_justification/fail_justification_enum_field.rs:25:26
   |
25 | #[derive(AiJsonTemplate, AiJsonTemplateWithJustification)]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: an implementation of `PartialEq` might be missing for `ContainsEnum`
  --> tests/trybuild_justification/fail_justification_enum_field.rs:28:1
   |
28 | struct ContainsEnum {
   | ^^^^^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `ContainsEnum` with `#[derive(PartialEq)]`
   |
28 + #[derive(PartialEq)]
29 | struct ContainsEnum {
   |

error[E0277]: the trait bound `BadEnum: ai_json_template::AiJsonTemplateWithJustification` is not satisfied
  --> tests/trybuild_justification/fail_justification_enum_field.rs:30:11
   |
30 |     data: BadEnum,
   |           ^^^^^^^ the trait `ai_json_template::AiJsonTemplateWithJustification` is not implemented for `BadEnum`
   |
   = help: the following other types implement trait `ai_json_template::AiJsonTemplateWithJustification`:
             ContainsEnum
             HashMap<K, V>
             Vec<T>
